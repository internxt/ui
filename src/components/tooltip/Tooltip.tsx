import { ReactNode, useRef, useState } from 'react';

export interface TooltipProps {
  children: ReactNode;
  title: string;
  subtitle?: string;
  popsFrom: 'right' | 'left' | 'top' | 'bottom';
  className?: string;
  delayInMs?: number;
  arrow?: boolean;
}

/**
 * Tooltip component
 *
 * @property {ReactNode} children
 * - The content that triggers the tooltip when hovered over.
 *
 * @property {string} title
 * - The main text displayed inside the tooltip. This is required.
 *
 * @property {string} [subtitle]
 * - An optional subtitle displayed below the main title inside the tooltip.
 *
 * @property {'right' | 'left' | 'top' | 'bottom'} popsFrom
 * - Determines the direction from which the tooltip appears relative to the trigger element.
 *   - "right": Tooltip appears to the right of the element.
 *   - "left": Tooltip appears to the left of the element.
 *   - "top": Tooltip appears above the element.
 *   - "bottom": Tooltip appears below the element.
 *
 * @property {string} [className]
 * - Additional CSS classes to style the tooltip container. Use to override default styles.
 *
 * @property {number} [delayInMs]
 * - The delay (in milliseconds) before hiding the tooltip after the mouse leaves the trigger element.
 *   - If not provided, the tooltip hides immediately.
 *
 * @property {boolean} [arrow=true]
 * - Whether to display the arrow pointing to the trigger element.
 *   - Default is true.
 *
 * @returns {JSX.Element}
 * - A tooltip component that shows additional information when hovering over its children.
 */

const Tooltip = ({
  children,
  title,
  subtitle,
  popsFrom,
  className,
  delayInMs,
  arrow = true,
}: TooltipProps): JSX.Element => {
  const [visible, setVisible] = useState(false);

  const timeoutRef = useRef<null | number>(null);

  function show() {
    setVisible(true);
  }

  function hide() {
    setVisible(false);
  }

  function handleMouseEnter() {
    if (timeoutRef.current !== null) {
      clearTimeout(timeoutRef.current);
    }
    show();
  }
  function handleMouseLeave() {
    if (delayInMs) {
      timeoutRef.current = setTimeout(() => {
        timeoutRef.current = null;
        hide();
      }, delayInMs) as unknown as number;
    } else {
      hide();
    }
  }

  let tooltipPosition = '';
  let trianglePosition = '';
  let triangle = '';

  switch (popsFrom) {
    case 'right':
      tooltipPosition = 'left-full top-1/2 -translate-y-1/2 ml-1.5';
      trianglePosition = 'flex-row-reverse';
      triangle = 'polygon(100% 0%, 100% 100%, 0% 50%)';
      break;
    case 'left':
      tooltipPosition = 'right-full top-1/2 -translate-y-1/2 mr-1.5';
      trianglePosition = 'flex-row';
      triangle = 'polygon(0% 0%, 0% 100%, 100% 50%)';
      break;
    case 'top':
      tooltipPosition = 'bottom-full left-1/2 -translate-x-1/2 mb-1.5 origin-bottom';
      trianglePosition = 'flex-col';
      triangle = 'polygon(0% 0%, 100% 0%, 50% 100%)';
      break;
    case 'bottom':
      tooltipPosition = 'top-full left-1/2 -translate-x-1/2 mt-1.5';
      trianglePosition = 'flex-col-reverse';
      triangle = 'polygon(50% 0%, 0% 100%, 100% 100%)';
      break;
  }

  return (
    <div
      className={`relative w-max ${className}`}
      onMouseEnter={handleMouseEnter}
      onMouseLeave={handleMouseLeave}
      style={{ lineHeight: 0 }}
    >
      <div
        className={`pointer-events-none absolute ${tooltipPosition} flex items-center ${trianglePosition} drop-shadow-tooltip transition-all duration-150 ${
          visible ? 'scale-100 opacity-100' : 'scale-50 opacity-0'
        }`}
      >
        <div className="w-max rounded-lg bg-gray-90 px-4 py-1.5 text-center dark:bg-gray-5">
          <p className="text-base text-white">{title}</p>
          {subtitle && <p className="-mt-1 text-sm text-gray-40">{subtitle}</p>}
        </div>
        {arrow && (
          <div
            className={`bg-gray-90 dark:bg-gray-5 ${
              popsFrom === 'bottom' || popsFrom === 'top' ? 'h-1.5 w-4' : 'h-4 w-1.5'
            }`}
            style={{ clipPath: triangle, marginTop: popsFrom === 'top' ? '-1px' : undefined }}
            data-testid="tooltip-arrow"
          />
        )}
      </div>
      {children}
    </div>
  );
};

export default Tooltip;
