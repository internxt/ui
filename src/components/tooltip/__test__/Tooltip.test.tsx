import { render, fireEvent, waitFor } from '@testing-library/react';
import Tooltip, { TooltipProps } from '../Tooltip';
import { expect } from 'chai';
import React, { Children } from 'react';
import { describe, afterEach, vi, it } from 'vitest';

describe('Tooltip', () => {
  const defaultProps: TooltipProps = {
    title: 'Tooltip Title',
    children: <button>Hover me</button>,
    popsFrom: 'top',
  };
  const renderLoader = (props = {}) => {
    return render(<Tooltip {...defaultProps} {...props} />);
  };

  afterEach(() => {
    vi.clearAllMocks();
  });

  it('should match snapshot', () => {
    const loader = renderLoader();
    expect(loader).toMatchSnapshot();
  });

  it('renders the children element', () => {
    const { getByText } = renderLoader();
    expect(getByText('Hover me')).toBeInTheDocument();
  });

  it('shows the tooltip on mouse enter', async () => {
    const { getByText } = renderLoader();

    const trigger = getByText('Hover me');
    fireEvent.mouseEnter(trigger);

    await waitFor(() => {
      expect(getByText('Tooltip Title')).toBeVisible();
    });
  });

  it('hides the tooltip on mouse leave without delay', async () => {
    const { getByText, queryByText } = renderLoader();

    const trigger = getByText('Hover me');
    fireEvent.mouseEnter(trigger);
    fireEvent.mouseLeave(trigger);

    const element = queryByText('Tooltip Title')?.parentElement?.parentElement;

    await waitFor(() => {
      expect(element).toHaveClass('opacity-0');
    });
  });

  it('applies delay before hiding tooltip when delayInMs is set', async () => {
    const { getByText, queryByText } = renderLoader({ delayInMs: 100 });

    const trigger = getByText('Hover me');
    fireEvent.mouseEnter(trigger);
    fireEvent.mouseLeave(trigger);

    const element = queryByText('Tooltip Title')?.parentElement?.parentElement;

    await waitFor(() => {
      expect(element).toHaveClass('opacity-100');
    });

    await waitFor(
      () => {
        expect(element).toHaveClass('opacity-0');
      },
      { timeout: 300 },
    );
  });

  it('renders subtitle if provided', () => {
    const { getByText } = renderLoader({ subtitle: 'Tooltip Subtitle' });

    fireEvent.mouseEnter(getByText('Hover me'));

    expect(getByText('Tooltip Subtitle')).toBeVisible();
  });

  it('renders the arrow by default', () => {
    const { getByText, getByTestId } = renderLoader();

    fireEvent.mouseEnter(getByText('Hover me'));
    const arrow = getByTestId('tooltip-arrow');

    expect(arrow).toBeInTheDocument();
  });

  it('does not render the arrow if `arrow` is false', () => {
    const { getByText, queryByTestId } = renderLoader({ arrow: false });

    fireEvent.mouseEnter(getByText('Hover me'));
    const arrow = queryByTestId('tooltip-arrow');

    expect(arrow).not.toBeInTheDocument();
  });

  it('applies the correct position when popsFrom is top', () => {
    const { getByText } = renderLoader();

    fireEvent.mouseEnter(getByText('Hover me'));

    const tooltip = getByText('Tooltip Title').parentElement?.parentElement;

    expect(tooltip).toHaveClass('bottom-full');
  });

  it('applies the correct position when popsFrom is bottom', () => {
    const { getByText } = renderLoader({ popsFrom: 'bottom' });

    fireEvent.mouseEnter(getByText('Hover me'));

    const tooltip = getByText('Tooltip Title').parentElement?.parentElement;

    expect(tooltip).toHaveClass('top-full');
  });

  it('applies the correct position when popsFrom is left', () => {
    const { getByText } = renderLoader({ popsFrom: 'left' });

    fireEvent.mouseEnter(getByText('Hover me'));

    const tooltip = getByText('Tooltip Title').parentElement?.parentElement;

    expect(tooltip).toHaveClass('right-full');
  });

  it('applies the correct position when popsFrom is right', () => {
    const { getByText } = renderLoader({ popsFrom: 'right' });

    fireEvent.mouseEnter(getByText('Hover me'));

    const tooltip = getByText('Tooltip Title').parentElement?.parentElement;

    expect(tooltip).toHaveClass('left-full');
  });

  it('applies custom className if provided', () => {
    const { getByText } = renderLoader({ className: 'custom-class' });

    const container = getByText('Hover me').parentElement;

    expect(container).toHaveClass('custom-class');
  });

  it('should clear the timeout on mouse enter', () => {
    const clearTimeoutMock = vi.fn();
    global.clearTimeout = clearTimeoutMock;

    const { getByText } = renderLoader({ delayInMs: 100 });
    fireEvent.mouseLeave(getByText('Hover me'));
    fireEvent.mouseEnter(getByText('Hover me'));

    expect(clearTimeoutMock).toHaveBeenCalled();
  });
});
