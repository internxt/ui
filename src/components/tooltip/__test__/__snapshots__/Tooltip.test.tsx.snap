// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tooltip > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="relative w-max undefined"
        style="line-height: 0;"
      >
        <div
          class="pointer-events-none absolute bottom-full left-1/2 -translate-x-1/2 mb-1.5 origin-bottom flex items-center flex-col drop-shadow-tooltip transition-all duration-150 scale-50 opacity-0"
        >
          <div
            class="w-max rounded-lg bg-gray-90 px-4 py-1.5 text-center dark:bg-gray-5"
          >
            <p
              class="text-base text-white"
            >
              Tooltip Title
            </p>
          </div>
          <div
            class="bg-gray-90 dark:bg-gray-5 h-1.5 w-4"
            data-testid="tooltip-arrow"
            style="clip-path: polygon(0% 0%, 100% 0%, 50% 100%); margin-top: -1px;"
          />
        </div>
        <button>
          Hover me
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="relative w-max undefined"
      style="line-height: 0;"
    >
      <div
        class="pointer-events-none absolute bottom-full left-1/2 -translate-x-1/2 mb-1.5 origin-bottom flex items-center flex-col drop-shadow-tooltip transition-all duration-150 scale-50 opacity-0"
      >
        <div
          class="w-max rounded-lg bg-gray-90 px-4 py-1.5 text-center dark:bg-gray-5"
        >
          <p
            class="text-base text-white"
          >
            Tooltip Title
          </p>
        </div>
        <div
          class="bg-gray-90 dark:bg-gray-5 h-1.5 w-4"
          data-testid="tooltip-arrow"
          style="clip-path: polygon(0% 0%, 100% 0%, 50% 100%); margin-top: -1px;"
        />
      </div>
      <button>
        Hover me
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
